Distribution/Server/Features/ReverseDependencies/State.hs

```diff
@@ -13,406 +15,203 @@ import Distribution.PackageDescription
 import Distribution.Version
 
 import Data.Acid (Query, Update, makeAcidic)
-import Data.List (foldl', union)
-import Data.Maybe (maybeToList, fromMaybe)
-import Data.SafeCopy (base, deriveSafeCopy)
+import Data.List (union, (\\))
+import Data.Maybe (maybeToList)
+import Data.SafeCopy hiding (Version)
 import Data.Typeable (Typeable)
 import Data.Map (Map)
 import qualified Data.Map as Map
+import Data.Bimap (Bimap, (!), (!>))
+import qualified Data.Bimap as Bimap
+import Control.Applicative
 import Data.Set (Set)
 import qualified Data.Set as Set
-
-import Data.STRef
+import Control.Monad (void)
 import Control.Monad.ST
 import Control.Monad.State (put, get)
-import Control.Monad.Reader (ask, asks)
-
--- The main reverse dependencies map is a drawn-out Map PackageId PackageId,
--- with an extra component to store ranges (all ranges that *could* be revdeps,
--- even if no packages in the index currently satisfy those ranges).
---
--- For selected entries of the map (foo, (2.0, (bar, [1.0]), (bar, (1.0, <3)))):
--- This means that bar-1.0 depends on a version of foo <3, and foo 2.0 meets this criterion.
-type RevDeps = Map PackageName (Map Version (Map PackageName (Set Version)), Map PackageName (Map Version VersionRange))
-
-type CombinedDeps = Map PackageName VersionRange
-
--- TODO: should this be (Maybe (Version, Maybe VersionStatus))?
--- it should be possible, albeit with a bit more work, to determine all revdeps
--- of a package which don't have any versions presently satisfying them.
--- (for an entry (a, b) of RevDeps, take (union a \ b))
-type ReverseDisplay = Map PackageName (Version, Maybe VersionStatus)
+import Control.Monad.Reader (ask)
+import Data.Graph (Graph, Vertex)
+import qualified Data.Graph as Gr
+import qualified Data.Array as Arr ((!), accum, bounds)
 
-data ReverseIndex = ReverseIndex {
-    -- this field is a duplication of PackageIndex, so updates don't have
-    -- to use much information outside of this component, resulting in huge
-    -- happstack-state event files
-    duplicatedIndex :: PackageIndex PackageId,
+type NodeId = Int
+type RevDeps = Graph
 
-    -- The main reverse dependencies map
-    reverseDependencies :: RevDeps,
+data ReverseIndex = ReverseIndex {
+    duplicatedIndex     :: !(PackageIndex PackageId),
```

Comment:
It looks like this genuinely is redundant. It is used during lookups, but these are cases where we could grab the `PackageIndex` from the core feature and use that in combination. This will be a tad easier if we switch to using a `MemState`. Then the query functions will just need to take the `ReverseIndex` and the core `PackageIndex` to do their stuff.

It's not small so it'd be better if we can eliminate it.